receiver/elasticsearchreceiver/integration_test.go

```diff
@@ -0,0 +1,97 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package elasticsearchreceiver
+
+import (
+	"context"
+	"net"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/scrapertest"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/scrapertest/golden"
+	"github.com/stretchr/testify/require"
+	"github.com/testcontainers/testcontainers-go"
+	"github.com/testcontainers/testcontainers-go/wait"
+
+	"go.opentelemetry.io/collector/component/componenttest"
+	"go.opentelemetry.io/collector/consumer/consumertest"
+)
+
+var (
+	containerRequest7_9_3 = testcontainers.ContainerRequest{
+		FromDockerfile: testcontainers.FromDockerfile{
+			Context:    filepath.Join("testdata", "integration"),
+			Dockerfile: "Dockerfile.elasticsearch.7_9_3",
+		},
+		ExposedPorts: []string{"9200:9200"},
+		WaitingFor: wait.ForListeningPort("9200").
+			WithStartupTimeout(2 * time.Minute),
+	}
+)
+
+func TestElasticsearchIntegration(t *testing.T) {
+	//Starts an elasticsearch docker container
+	t.Run("Running elasticsearch 7.9", func(t *testing.T) {
+		t.Parallel()
+		container := getContainer(t, containerRequest7_9_3)
+		defer func() {
+			require.NoError(t, container.Terminate(context.Background()))
+		}()
+		//todo
+		hostname, err := container.Host(context.Background())
+		require.NoError(t, err)
+
+		f := NewFactory()
+		cfg := f.CreateDefaultConfig().(*Config)
+		cfg.Endpoint = net.JoinHostPort(hostname, "9200")
+
+		consumer := new(consumertest.MetricsSink)
+		settings := componenttest.NewNopReceiverCreateSettings()
+		rcvr, err := f.CreateMetricsReceiver(context.Background(), settings, cfg, consumer)
+		require.NoError(t, err, "failed creating metrics receiver")
+		require.NoError(t, rcvr.Start(context.Background(), componenttest.NewNopHost()))
+		require.Eventuallyf(t, func() bool {
+			return len(consumer.AllMetrics()) > 0
+		}, 2*time.Minute, 1*time.Second, "failed to receive more than 0 metrics")
+		require.NoError(t, rcvr.Shutdown(context.Background()))
+
+		actualMtrics := consumer.AllMetrics()[0]
+
+		expectedFile := filepath.Join("testdata", "integration", "expected.7_9.json")
+		expectedMetrics, err := golden.ReadMetrics(expectedFile)
+		require.NoError(t, err)
+
+		scrapertest.CompareMetrics(expectedMetrics, actualMtrics, scrapertest.IgnoreMetricValues())
+		//err = scrapertest.CompareMetrics(expectedMetrics, actualMtrics, scrapertest.IgnoreMetricValues())
```

Comment:
Were you able to run this receiver locally against the container? (Is that how you generated the result?)

Reply:
yes, I run this reveiver locally and get the result. I have checked the es container by call `http://localhost:9200`,its seem running well.

Reply:
If it's working for you locally, then I suspect this is an issue with not using a "wait strategy" to verify that the application is ready. Without a wait strategy, the container may start, but the application within it may not be ready to respond by the time the scraper tries to query it.

[Here's an example of a typical wait strategy.](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/14773563ba0ecf5dd23f4985f7c14574d77c0a7f/receiver/mysqlreceiver/integration_test.go#L114)

[Here's an example of a custom wait strategy.](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/14773563ba0ecf5dd23f4985f7c14574d77c0a7f/receiver/apachereceiver/integration_test.go#L96)

Reply:
thx ,I will have a try...